File: Day48/Part4/arcane_forge.cpp
#include <iostream>
#include <vector>

using namespace std;

struct Machine {
    long long craft;
    long long cd;
    long long value;
};

long long solveArcaneForge(long long t, const vector<Machine>& machines) {
    long long total_value = 0;

    for (const auto& m : machines) {
        long long cycle_time = m.craft + m.cd;
        long long count = t / cycle_time;
        long long rem = t % cycle_time;

        if (rem >= m.craft) {
            count++;
        }

        total_value += count * m.value;
    }

    return total_value;
}

int main() {
    long long T = 10;
    vector<Machine> Machines = {
        {2, 1, 5},
        {5, 2, 20}
    };

    long long result = solveArcaneForge(T, Machines);
    cout << "Maximum Value: " << result << endl;

    return 0;
}

File: Day48/Part4/arcane_forge.py
def solve_arcane_forge(t, machines):
    total_value = 0
    
    for craft, cd, value in machines:
        cycle_time = craft + cd
        
        # Full cycles
        count = t // cycle_time
        
        # Remaining time
        rem = t % cycle_time
        
        # Can we finish one last item?
        # We only need 'craft' time to finish the item.
        # The cooldown happens after, but if T ends during cooldown, the item is already done.
        if rem >= craft:
            count += 1
            
        total_value += count * value
        
    return total_value

if __name__ == "__main__":
    T = 10
    # craft, cd, value
    Machines = [
        [2, 1, 5],
        [5, 2, 20]
    ]
    
    result = solve_arcane_forge(T, Machines)
    print(f"Maximum Value: {result}")
